alternator/conditions.cc

```diff
@@ -179,6 +181,59 @@ static bool check_BEGINS_WITH(const rjson::value* v1, const rjson::value& v2) {
     return val1.substr(0, val2.size()) == val2;
 }
 
+static std::string_view get_strv(const rjson::value& v) {
+    return std::string_view(v.GetString(), v.GetStringLength());
+}
+
+static bool is_set_of(const rjson::value& type1, const rjson::value& type2) {
+    return (type2 == "S" && type1 == "SS") || (type2 == "N" && type1 == "NS") || (type2 == "B" && type1 == "BS");
+}
+
+// Check if two JSON-encoded values match with the CONTAINS relation
+static bool check_CONTAINS(const rjson::value* v1, const rjson::value& v2) {
+    if (!v1) {
+        return false;
+    }
+    const auto& kv1 = *v1->MemberBegin();
+    const auto& kv2 = *v2.MemberBegin();
+    if (kv2.name != "S" && kv2.name != "N" &&  kv2.name != "B") {
+        throw api_error("ValidationException",
+                        format("CONTAINS operator requires a single AttributeValue of type String, Number, or Binary, "
+                               "got {} instead", kv2.name));
+    }
+    if (kv1.name == "S" && kv2.name == "S") {
+        return get_strv(kv1.value).find(get_strv(kv2.value)) != std::string_view::npos;
+    } else if (kv1.name == "B" && kv2.name == "B") {
+        return base64_decode(kv1.value).find(base64_decode(kv2.value)) != bytes::npos;
```

Comment:
Unfortunately, the encoded sequences somehow do not match, and the test fails when I unify the cases:
```
        test_table_s.update_item(Key={'p': p},
            AttributeUpdates={'z': {'Value': 5, 'Action': 'PUT'}},
>           Expected={'d': {'ComparisonOperator': 'CONTAINS', 'AttributeValueList': [bytearray('here', 'utf-8')]}}
        )
[snip]
E           botocore.errorfactory.ConditionalCheckFailedException: An error occurred (ConditionalCheckFailedException) when calling the UpdateItem operation: Failed condition.
```

In the debugger, I can see the following payloads:
```
(gdb) p kv1.value
$33 = {data_ = {s = {length = 1735083873, hashcode = 1818773348, str = 0x1c0501003d556d63 <error: Cannot access memory at address 0x1c0501003d556d63>}, ss = {str = "aGkgdGhlcmU=\000\001"}, n = {i = {i = 1735083873, padding = "dGhl"}, u = {u = 1735083873, padding2 = "dGhl"}, i64 = 7811572050231510881, u64 = 7811572050231510881, d = 1.6346903834732492e+214}, o = {size = 1735083873, capacity = 1818773348, members = 0x1c0501003d556d63}, a = {size = 1735083873, capacity = 1818773348, elements = 0x1c0501003d556d63}, f = {payload = "aGkgdGhlcmU=\000\001", flags = 7173}}}
(gdb) p kv2.value
$34 = {data_ = {s = {length = 2035697505, hashcode = 1027428698, str = 0x1c05050000000000 <error: Cannot access memory at address 0x1c05050000000000>}, ss = {str = "aGVyZQ==\000\000\000\000\000\005"}, n = {i = {i = 2035697505, padding = "ZQ=="}, u = {u = 2035697505, padding2 = "ZQ=="}, i64 = 4412772658917558113, u64 = 4412772658917558113, d = 1.0415770210153851e-13}, o = {size = 2035697505, capacity = 1027428698, members = 0x1c05050000000000}, a = {size = 2035697505, capacity = 1027428698, elements = 0x1c05050000000000}, f = {payload = "aGVyZQ==\000\000\000\000\000\005", flags = 7173}}}
```

The unified code looks like you would expect:
```c++
    if ((kv1.name == "S" && kv2.name == "S") || (kv1.name == "B" && kv2.name == "B")) {
        return get_strv(kv1.value).find(get_strv(kv2.value)) != std::string_view::npos;
    }
```

I'm not really sure why this doesn't work but BEGINS_WITH does. :man_shrugging:

Reply:
Base64 encodes each 3 input bytes as 4 bytes (each input 6 bits as output 1 byte). If we look at the fifth input byte - as our test does, its encoding doesn't start at a byte boundary. So doing a byte-string match would NOT work. Fully decoding the string is the easiest implementation but not the quickest - it should be possible to implement a find() algorithm in an encoded string while decoding it on-the-fly - without actually opening it into a new string. But let's not do it now.
By the way, BEGINS_WITH also shouldn't work this way! The beginning of the match check indeed starts on a byte boundary, but it doesn't have to end on one. I'll check now if our test is not exhaustive enough and missing a bug in this.

Reply:
Yes, this is a real bug in BEGINS_WITH, and indeed the test missed it because it forgot to test the byte-array case :-( I opened this issue: https://github.com/scylladb/scylla/issues/5457

Reply:
Oh... it was stuck in my head for some reason that we're just doing full string comparisons here, not CONTAINS. Makes sense now.